### PR TITLE
Disconnect from gossiping peer

### PR DESCRIPTION
**Pull request by jBarz**
- Add a disconnect command
  `lighting-cli disconnect <peer id>`
- Add test for disconnect command

### CODE REVIEW

File: tests/test_lightningd.py [hunk 1]
```diff
@@ -3802,6 +3802,35 @@ def test_blockchaintrack(self):
         # Our funds got reorged out, we should not have any funds that are confirmed
         assert [o for o in l1.rpc.listfunds()['outputs'] if o['status'] != "unconfirmed"] == []
 
+    def test_disconnect(self):
```
Comment (ZmnSCPxj) [inline tests/test_lightningd.py]: There is an existing `test_disconnect` in line 2408 (CI reported this)

File: gossipd/gossip.c [hunk 2]
```diff
@@ -1011,6 +1011,32 @@ static struct io_plan *hand_back_peer(struct io_conn *conn,
...
+	peer = find_peer(daemon, &id);
+	if (peer && peer->local) {
+		io_close(peer->local->conn);
+		msg = towire_gossipctl_peer_disconnect_reply(msg);
+		daemon_conn_send(&daemon->master, take(msg));
+	} else {
+		status_trace("release_peer: peer %s %s",
+			     type_to_string(trc, struct pubkey, &id),
+			     !peer ? "not found"
+			     : peer->local ? "already releasing"
```
Comment (ZmnSCPxj) [inline gossipd/gossip.c]: I think this particular case is impossible given the `if` condition?  If `peer` is non-NULL and `peer->local` is set then we would have gotten into the "then" branch of the `if` above.  Ignore this comment if this is for future-proofness though...

(A comment to the effect that "local peers are gossiping, non-local peers are owned by another daemon and are not gossiping" would be good too...)

File: gossipd/gossip.c [hunk 3]
```diff
@@ -1011,6 +1011,32 @@ static struct io_plan *hand_back_peer(struct io_conn *conn,
...
+	if (!fromwire_gossipctl_peer_disconnect(msg, &id))
+		master_badmsg(WIRE_GOSSIPCTL_PEER_DISCONNECT, msg);
+
+	peer = find_peer(daemon, &id);
+	if (peer && peer->local) {
+		io_close(peer->local->conn);
+		msg = towire_gossipctl_peer_disconnect_reply(msg);
+		daemon_conn_send(&daemon->master, take(msg));
+	} else {
+		status_trace("release_peer: peer %s %s",
```
Comment (ZmnSCPxj) [inline gossipd/gossip.c]: Function is named `disconnect_peer`...

Comment (jBarz) [reply gossipd/gossip.c]: I will rename the disconnect handler function to `peer_disconnect`

File: gossipd/gossip.c (hunk 2 above)
Comment (jBarz) [reply gossipd/gossip.c]: I will remove that additional redundant check

File: gossipd/gossip.c (hunk 3 above)
Comment (ZmnSCPxj) [reply gossipd/gossip.c]: `peer_disconnect` sounds more like "the peer disconnects from me", while `disconnect_peer` sounds more like "I disconnect from the peer", so I think `disconnect_peer` for this function is good.

Comment (ZmnSCPxj) [reply gossipd/gossip.c]: What I am commenting on here is that the status message starts with `release_peer:` when the function that prints the message is named `disconnect_peer`.

Comment (jBarz) [reply gossipd/gossip.c]: ahh yes. That makes sense. Will change.

**Comment by cdecker** [comment]
ACK af4a81c

File: lightningd/peer_control.c [hunk 4]
```diff
@@ -875,6 +875,55 @@ void activate_peers(struct lightningd *ld)
...
+/* Peer has been released from gossip. */
+static void gossip_peer_disconnected (struct subd *gossip,
+				 const u8 *resp,
+				 const int *fds,
+				 struct command *cmd) {
+	if (!fromwire_gossipctl_peer_disconnect_reply(resp)) {
+		if (!fromwire_gossipctl_peer_disconnect_replyfail(resp))
+			fatal("Gossip daemon gave invalid reply %s",
+			      tal_hex(gossip, resp));
+		command_fail(cmd, "Peer not connected");
```
Comment (ZmnSCPxj) [inline lightningd/peer_control.c]: There are actually 2 possible errors here, either peer is not connected, or peer has a channel and is doing something other than gossiping.  We can add a `bool` to `gossipctl_peer_disconnect_replyfail` to differentiate.

File: gossipd/gossip.c [hunk 5]
```diff
@@ -1011,6 +1011,31 @@ static struct io_plan *hand_back_peer(struct io_conn *conn,
...
+	peer = find_peer(daemon, &id);
+	if (peer && peer->local) {
+		/* This peer is local to this (gossipd) dameon */
+		io_close(peer->local->conn);
+		msg = towire_gossipctl_peer_disconnect_reply(msg);
+		daemon_conn_send(&daemon->master, take(msg));
+	} else {
+		status_trace("disconnect_peer: peer %s %s",
+			     type_to_string(trc, struct pubkey, &id),
+			     !peer ? "not found" : "not local to this daemon");
```
Comment (ZmnSCPxj) [inline gossipd/gossip.c]: `not local to this daemon` (aka gossipd) is equivalent to `not gossiping`, and may be clearer.

**Comment by ZmnSCPxj** [comment]
ACK d00862334d97d81148d6ffbe5b67b4ceb882bc60

**Pull request merged**